src/s3path.jl

```diff
@@ -0,0 +1,235 @@
+    )
+end
+
+function S3Path(str::AbstractString; config::AWSConfig=aws_config())
+    str = String(str)
+    root = ""
+    path = ()
+    dir = true
+
+    @assert startswith(str, "s3://")
+    tokenized = split(str, "/")
+    bucket = strip(tokenized[3], '/')
+    drive = "s3://$bucket"
+
+    if length(tokenized) > 3
+        root = "/"
+        # If the last tokenized element is an empty string then we've parsed a directory
+        dir = isempty(last(tokenized))
+        path = tuple(filter!(!isempty, tokenized[4:end])...)
+    end
+
+    return S3Path(path, root, drive, dir, config)
+end
+
+Base.print(io::IO, fp::S3Path) = print(io, fp.anchor * fp.key)
+
+function Base.:(==)(a::S3Path, b::S3Path)
+    return a.segments == b.segments &&
+        a.root == b.root &&
+        a.drive == b.drive
+end
+
+function Base.getproperty(fp::S3Path, attr::Symbol)
+    if isdefined(fp, attr)
+        return getfield(fp, attr)
+    elseif attr === :anchor
+        return fp.drive * fp.root
+    elseif attr === :separator
+        return "/"
+    elseif attr === :bucket
+        return split(fp.drive, "//")[2]
+    elseif attr === :key
+        return fp.dir ? join(fp.segments, '/') * "/" : join(fp.segments, '/')
+    else
+        # Call getfield even though we know it'll error
+        # so the message is consistent.
+        return getfield(fp, attr)
+    end
+end
+
+# We need to special case join and parents so that we propagate
+# directories correctly
+function Base.join(prefix::S3Path, pieces::AbstractString...)
+    segments = String[prefix.segments...]
+    dir = endswith(last(pieces), "/")
+
+    for p in pieces
+        push!(segments, filter!(!isempty, split(p, "/"))...)
+    end
+
+    return S3Path(
+        tuple(segments...),
+        prefix.root,
+        prefix.drive,
+        dir,
+        prefix.config,
+    )
+end
+
+function parents(fp::S3Path)
+    if hasparent(fp)
+        return map(1:length(fp.segments)-1) do i
+            S3Path(fp.segments[1:i], fp.root, fp.drive, true, fp.config)
+        end
+    else
+        error("$fp has no parents")
+    end
+end
+
+FilePathsBase.ispathtype(::Type{S3Path}, str::AbstractString) = startswith(str, "s3://")
+FilePathsBase.isabs(fp::S3Path) = !isempty(fp.drive) && !isempty(fp.root)
+FilePathsBase.expanduser(fp::S3Path) = fp
+FilePathsBase.exists(fp::S3Path) = s3_exists(fp.config, fp.bucket, fp.key)
+Base.isfile(fp::S3Path) = !fp.dir && exists(fp)
+function FilePathsBase.isdir(fp::S3Path)
+    # Note: objects "s3://bucket/a" and "s3://bucket/a/b" can co-exist. If both of these
+    # objects exist listing the keys for "s3://bucket/a" returns ["s3://bucket/a"] while
+    # "s3://bucket/a/" returns ["s3://bucket/a/b"].
+    if isempty(fp.segments)
+        key = ""
+    elseif fp.dir
+        key = fp.key
+    else
+        return false
+    end
+
+    objects = s3_list_objects(fp.config, fp.bucket, key; max_items=1)
+    return !isempty(objects)
+end
+
+Base.real(fp::S3Path) = fp
```

Comment:
Can we comment this with reference to the FilePathBase docs?

Reply:
Oops, we don't have pushed docs setup for FilePathsBase yet.